### ISSUE TITLE
[DCR]: Extracted files should use filesystem read-only attribute

### ISSUE THREAD

**Issue by stkenned-havok** [NuGet/Home#11286]
### NuGet Product(s) Affected

NuGet.exe

### Current Behavior

Immutability is highly desirable for reliable package management. Many parts of the chain already support this e.g. azure artifacts & package signing. The last stage is on the developers machine. Ideally, we'd be able to mount a read-only zip directly in the filesystem. But until that happens, we should make use of the OS facilities which are available - e.g. the humble read-only permission bit.

### Desired Behavior

This has been requested before in #1729 and #10404. We got a partial solution of teaching some editors to pretend the files are read-only. But there are plenty of other ways to accidentally modify files. Let's make that a little harder by making the files read-only.

**Comment by stkenned-havok** [comment]
Perhaps there are some other considerations(?), but this looks trivial to implement. In src\NuGet.Core\NuGet.Packaging\PackageExtraction\ZipArchiveExtensions.cs we're already updating the lastwritetime attribute, so could also set the flag there.

**Comment by jehumb-havok** [comment]
Strongly agreed. I changed some NuGet package file by mistake from a locally-built package, instead of the actual sources. I almost lost 2 hours of work because of that, as I caught at the last minute that the wrong files were modified, just before the package was about to be overwritten with a new version. This would never have happened with read-only files.

I note #1729 was never looked at. #10404 was dismissed as a breaking change, but I have yet to hear about an argument where NuGet package files are used with write access in a valid scenario. And even in that case, making a local copy elsewhere sounds like an easy workaround. I'll add some other dev tools like Unity also make their packages content read-only.

**Comment by txwizard** [comment]
I strongly agree, especially given that the NuGet installer never overwrites a package. One thing, however, that must be taken into account to avoid a break is that reinstallation must clear the RO flag if it is set. Nevertheless, this is again a trivial change.

**Comment by zivkan** [comment]
`dotnet nuget locals global-packages --clear` will probably need to be fixed. It's probably the same code path for `nuget.exe locals global-packages -clear` and Visual Studio's clear button, but if not, those would need to be fixed as well. Packages.config also needs to be tested, as when a package version is no longer used by any packages.config in the solution, NuGet deletes it from the solution packages folder.

I did a quick test, and building a project when the global packages folder is read only does not copy the read-only flag. Therefore, there's no problem if you change the version of the package and build again, the different assembly version will be overwrite what's already in the bin folder.

Those are the only concerns I can think of in short notice, but I like the idea.

**Comment by zivkan** [comment]
Actually, our test infrastructure might need updating too, as many tests use a test-specific global packages folder, which is cleaned up at the end of the test.

**Comment by stkenned-havok** [comment]
Can an admin reopen this? @zivkan?

Community has asked several times for this feature #1729 and #10404 and #11286, it's not stale.
